README.md
[中文](README_zh_CN.md)

# siyuan-plugin-local-service

<img src="./icon.png" width="160" height="160" alt="icon">

a plugin for connecting siyuan-note with local services

> View all commands: `windos.zhi`

## Quick start - TODO

  ```js
await zhi.npm.checkAndInitNode()
zhi.app.store.rubick.bootstrap()
await zhi.app.store.rubick.getPackage("python-hello")
await zhi.app.store.rubick.getPackage("local-service-chatgpt")
await zhi.app.store.rubick.getPackage("local-service-nocodb")
zhi.sc.stopAll()
zhi.sc.startAll()
  ```

  Open browser access, or set the URL to the WebApp plugin

  ```
  http://localhost:3000
  http://localhost:8000
  ```

```bash
> zhi
{
  app: 
    pd: {downloadAndExtractPackage},
  cmd: CustomCmd {},
  common: {StrUtil: ƒ, JsonUtil: ƒ, DateUtil: ƒ},
  device: class,
  if: class InvokeFactory,
  logger: (n2, $2, p) => {...},
  npm: NpmPackageManager {...},
  sc: ServiceManager {...}
  status: {serviceInited: true, deviceInited: true, cmdInited: true, infraInited: true},
}
```

* Unnecessary `node_modules` dependencies have been removed.

* Added Node download and npm dependency installation via `zhi.npm.checkAndInitNode()`.

* Added a service manager and mount it to `zhi.sc`

```
zhi.sc.findAll()
zhi.sc.findByServiceName(serviceName)
zhi.sc.startAll()
zhi.sc.startMany(services)
zhi.sc.stopAll()
zhi.sc.startByServiceName(serviceName)
zhi.sc.stopByServiceName(serviceName)
```

* Add a service package downloader and mount it to `zhi.app.pd`

```js
await zhi.app.pd.downloadAndExtractPackage(
  "https://ghproxy.com/https://github.com/terwer/siyuan-plugin-publisher/releases/download/v1.17.3/package.zip",
  "/Users/terwer/Downloads"
)
```

- Add service call factory InvokeFactory and mount to `zhi.if`

```js
const nodeInvoke = zhi.if.createInvoke("node")
const args = [
  "https://ghproxy.com/https://github.com/terwer/siyuan-plugin-publisher/releases/download/v1.17.3/package.zip",
  "/Users/terwer/Downloads",
]
await nodeInvoke.invoke("package-downloader", "services/package-downloader/downloadAndExtractPackage.cjs", args)
```

```js
// python invokde
const pythonInvode = zhi.if.createInvoke("python")
const args = []
await pythonInvode.invoke("python-hello", "services/python-hello/hello.py", args)
```

## Command List

* Download and install Node

```js
await zhi.npm.checkAndInitNode()
```

Initialization process:

```js
await zhi.npm.checkAndInitNode()
// [zhi] [2023-10-21 14:42:36] [INFO] [npm-package-manager] Node environment doesn't exist, preparing to install Node...
// Executing command: [thisPluginBasePath]/setup.cjs, args => v18.18.2, [appSiyuancommunityFolder]/node, options => {cwd: '[appSiyuancommunityFolder]/node', silent: true}
// Command execution log has been saved to the file => /Users/terwer/electron-command-log.txt
// [zhi] [2023-10-21 14:42:42] [INFO] [npm-package-manager] Node installation successful! 😄
// [zhi] [2023-10-21 14:42:42] [INFO] [package-helper] Dependencies updated successfully at [appSiyuancommunityFolder]/workspace/test/package.json
// [zhi] [2023-10-21 14:42:42] [INFO] [npm-package-manager] Detected deps.json change. Will install node_module once if needed, please wait...
// [zhi] [2023-10-21 14:42:42] [INFO] [npm-package-manager] npmCmd options => {cwd: '[appSiyuancommunityFolder]/workspace/test', env: {…}}
// [zhi] [2023-10-21 14:42:50] [INFO] [npm-package-manager] All node_module installed successfully
// [zhi] [2023-10-21 14:42:50] [INFO] [npm-package-manager] Package hash updated successfully
true
```

If already initialized, it's ignored:

```js
await zhi.npm.checkAndInitNode()
// [zhi] [2023-10-21 14:48:16] [INFO] [npm-package-manager] Node already installed, ignore
// [zhi] [2023-10-21 14:48:16] [INFO] [package-helper] deps.json hasn't changed since the last update, skip
true
```

* Test Node and npm

Test `node`:

```js
await zhi.npm.nodeCmd("-v")
// 'v18.18.2'
await zhi.npm.nodeVersion()
// 'v18.18.2'
await zhi.npm.electronNpmVersion()
// '18.15.0'
await zhi.npm.systemNpmVersion()
// 'v18.16.0'
```

Or:

```js
const path = require("path")
const thisPluginBasePath = path.join(window.siyuan.config.system.dataDir, "plugins", "siyuan-plugin-local-service")
await zhi.npm.nodeCmd(`${thisPluginBasePath}/hello.js`)
// 'Hello, World!'
```

Test `npm`:

```js
await zhi.npm.npmCmd("-v")
// '9.8.1'
await zhi.npm.npmVersion()
// '9.8.1'
```

* Other Commands

Install a dependency:

```js
// Cannot require if not installed
require("vue")
// node:internal/modules/cjs/loader:1085 Uncaught Error: Cannot find module 'vue'
```

Install:

```js
await zhi.npm.npmInstall("vue")
```

Install and require:

```js
const vue = await zhi.npm.requireInstall("vue")
```

Use Electron's built-in Node to execute commands:

```js
const path = require("path")
const thisPluginBasePath = path.join(window.siyuan.config.system.dataDir, "plugins", "siyuan-plugin-local-service")
// Point to the .js file you want to run
const command = `${thisPluginBasePath}/hello.js`
const args = []
const cwd = undefined
const result = await zhi.cmd.executeCommandWithBundledNodeAsync(command, args, cwd)
```

Use locally installed Node to execute commands:

```js
const path = require("path")
const thisPluginBasePath = path.join(window.siyuan.config.system.dataDir, "plugins", "siyuan-plugin-local-service")
await zhi.npm.nodeCmd(`${thisPluginBasePath}/hello.js`)
```

Use the system's Node to execute commands:

```js
const path = require("path")
const thisPluginBasePath = path.join(window.siyuan.config.system.dataDir, "plugins", "siyuan-plugin-local-service")
// Point to the .js file you want to run
const command = `${thisPluginBasePath}/hello.js`
const args = []
const options = undefined
await zhi.cmd.executeCommand("node", [`${command}`], options)
```

Use the system's Python to execute commands:

```js
await zhi.cmd.executeCommand("python", ["-V"])
// 'Python 3.11.4'
```

* Service command

```js
await zhi.pd.downloadAndExtractPackage("https://ghproxy.com/https://github.com/terwer/siyuan-plugin-publisher/releases/download/v1.17.3/package.zip")
await zhi.pd.downloadAndExtractPackage(
    "https://ghproxy.com/https://github.com/terwer/siyuan-plugin-publisher/releases/download/v1.17.3/package.zip",
    "/Users/terwer/Downloads"
)
```

or

```js
const path = require("path")
const thisPluginBasePath = path.join(window.siyuan.config.system.dataDir, "plugins", "siyuan-plugin-local-service")
const command = `${thisPluginBasePath}/services/package-downloader/index.cjs`
await zhi.npm.nodeCmd(command)
```

or factory mode call

```js
const nodeInvoke = zhi.if.createInvoke("node")
const args = [
  "https://ghproxy.com/https://github.com/terwer/siyuan-plugin-publisher/releases/download/v1.17.3/package.zip",
  "/Users/terwer/Downloads",
]
await nodeInvoke.invoke("package-downloader", "services/package-downloader/downloadAndExtractPackage.cjs", args)
```

```js
// python invokde
const pythonInvode = zhi.if.createInvoke("python")
const args = []
await pythonInvode.invoke("python-hello", "services/python-hello/hello.py", args)
```

## Explanation of Service Parameters

```js
const initParams = ["[thisPluginBasePath]", true]

const initParams = [
  "PORT[eq]8888",
  "NC_DB[eq]sqlite3:///?database=[siyuanDataDir]/storage/services/[thisServiceName]/noco.db",
]
```

Available placeholders

- `[thisPluginBasePath]` - The root directory of the local service plugin, for example: /Users/terwer/Documents/mydocs/SiYuanWorkspace/test/data/plugins/siyuan-plugin-local-service

- `[eq]` - The "=" symbol, used to separate key-value type parameters

- `[siyuanDataDir]` - The data directory of SiYuan Notes, for example: /Users/terwer/Documents/mydocs/SiYuanWorkspace/test/data

- `[thisServiceBasePath]` - The root directory of the current service, for example: /Users/terwer/Library/Application Support/siyuancommunity/workspace/test/apps/local-service-nocodb

- `[thisServiceName]` - The root directory of the current service, for example: local-service-nocodb

Example:

```js
const nodeInvoke = zhi.if.createInvoke("node")
const args = [
  "PORT=8888",
  "NC_DB=sqlite3:///?database[eq][siyuanDataDir]/storage/services/[thisServiceName]/noco.db",
]
await nodeInvoke.invoke("local-service-nocodb", "local-service-nocodb/index.js", args)
```

## Thanks

[rubick](https://github.com/rubickCenter/rubick)

CHANGELOG.md
# Changelog

## [1.4.0](https://github.com/terwer/siyuan-plugin-local-service/compare/v1.3.0...v1.4.0) (2023-10-24)
### Features
* 支持自定义执行目录和环境变量 ([123a647](https://github.com/terwer/siyuan-plugin-local-service/commit/123a6473f868d4f21d789432870521478eb2782c))
* 核心服务默认启动 ([dacc223](https://github.com/terwer/siyuan-plugin-local-service/commit/dacc223ac2541c581f90603344cc289a6f299093))
### Bug Fixes
* **deps:** bump actions/setup-node from 3 to 4 ([d54d0a9](https://github.com/terwer/siyuan-plugin-local-service/commit/d54d0a9b9f9983d10a37887d5e01df00787f1a26))
## [1.3.0](https://github.com/terwer/siyuan-plugin-local-service/compare/v1.2.0...v1.3.0) (2023-10-23)
* 优化启动与加载顺序 ([b92e156](https://github.com/terwer/siyuan-plugin-local-service/commit/b92e1565d759a4218983bbf585f34040a5ad07fe))
* 适配 chatgpt 和 nocodb ([09b113d](https://github.com/terwer/siyuan-plugin-local-service/commit/09b113d2621f32d21e730d8828dbd047a785bd73))
* 修复路径问题 ([77225e2](https://github.com/terwer/siyuan-plugin-local-service/commit/77225e204c198c83ef3171d17c9a60687f17df4e))
## [1.2.0](https://github.com/terwer/siyuan-plugin-local-service/compare/v1.1.0...v1.2.0) (2023-10-22)
* 支持 node 和 python 调用 ([8995b31](https://github.com/terwer/siyuan-plugin-local-service/commit/8995b31ac0edc838cb54461ba853cc37d00f54a0))
* 新增包下载器 ([15ef089](https://github.com/terwer/siyuan-plugin-local-service/commit/15ef089c0c4cf4282479a5c356f505a5d3db7fcf))
* 新增包下载器-下载与缓存服务包 ([1703cd9](https://github.com/terwer/siyuan-plugin-local-service/commit/1703cd98574d186b5a3e75c59b2227ba48d96722))
* 新增服务管理器 ([ed437ea](https://github.com/terwer/siyuan-plugin-local-service/commit/ed437ea350d167d872046aac226820a5207540b2))
## [1.1.0](https://github.com/terwer/siyuan-plugin-local-service/compare/v1.0.3...v1.1.0) (2023-10-21)
* 初始化安装node以及预定义依赖 ([62d0e8a](https://github.com/terwer/siyuan-plugin-local-service/commit/62d0e8a31845ae7e97ad208a8b4595a40462c29f))
* 统一规范化管理命令 ([6de6a22](https://github.com/terwer/siyuan-plugin-local-service/commit/6de6a22a6583128e6cedf7f9e41402cffcd378ce))
* 重构基础设施 ([f90f2ff](https://github.com/terwer/siyuan-plugin-local-service/commit/f90f2ff4147b4af10299894e207498c044aa33e5))
* **deps:** bump zhi-device from 2.6.0 to 2.8.1 ([44c0962](https://github.com/terwer/siyuan-plugin-local-service/commit/44c09623aa76fd24f1269ab89e0c3969f561e9f4))
### Miscellaneous
* **deps-dev:** bump @types/minimist from 1.2.3 to 1.2.4 ([a5a405b](https://github.com/terwer/siyuan-plugin-local-service/commit/a5a405b2bf9def33533d03d51f5cbc4e7e716309))
* **deps-dev:** bump svelte from 4.2.1 to 4.2.2 ([0ab7887](https://github.com/terwer/siyuan-plugin-local-service/commit/0ab788723f2d0a4c223d1b306e4e01c68295d689))
* **deps-dev:** bump vite from 4.4.11 to 4.5.0 ([3326679](https://github.com/terwer/siyuan-plugin-local-service/commit/33266793607a299dc26eb26ca62bc87a955cd488))
## [1.0.3](https://github.com/terwer/siyuan-plugin-local-service/compare/v1.0.2...v1.0.3) (2023-10-18)
* 修改预览图标 ([d3d36e5](https://github.com/terwer/siyuan-plugin-local-service/commit/d3d36e50fb481170827647d1a36e6a3d439b8f12))
* **deps-dev:** bump siyuan from 0.8.5 to 0.8.6 ([359713a](https://github.com/terwer/siyuan-plugin-local-service/commit/359713aeec743d1534305760bbd71eb988726475))
## [1.0.2](https://github.com/terwer/siyuan-plugin-local-service/compare/v1.0.1...v1.0.2) (2023-10-17)
* 修复缺失的 cross-spawn 包 ([39bce9e](https://github.com/terwer/siyuan-plugin-local-service/commit/39bce9e40a93eb800a699528d9ef844b5f7e6515))
## [1.0.1](https://github.com/terwer/siyuan-plugin-local-service/compare/v1.0.0...v1.0.1) (2023-10-17)
* 更新开发计划 ([88ee9f6](https://github.com/terwer/siyuan-plugin-local-service/commit/88ee9f6f8174f31a606f816533cb8f033522916f))
## 1.0.0 (2023-10-16)
* 初始化本地服务插件 ([5c4ac5f](https://github.com/terwer/siyuan-plugin-local-service/commit/5c4ac5fe014f659d0968b93b478c07ca8b756246))
* 初步完成核心加载流程 ([976cce8](https://github.com/terwer/siyuan-plugin-local-service/commit/976cce8566188a5a48e33ab2db83e8a561b8ff69))
* 基本启动流程 ([36a66d1](https://github.com/terwer/siyuan-plugin-local-service/commit/36a66d11252cb435782f1b42aaf8c9778b92b40d))
* 整合基础设施、命令工具 ([3955bbc](https://github.com/terwer/siyuan-plugin-local-service/commit/3955bbc53a22a1deae209a8485b1e90c23ad2aba))
* **deps:** bump actions/checkout from 3 to 4 ([9604c83](https://github.com/terwer/siyuan-plugin-local-service/commit/9604c83eacfdd68bd86362d5c2b7e37b1f9710f0))
* **deps:** bump zhi-lib-base from 0.2.6 to 0.5.0 ([5251393](https://github.com/terwer/siyuan-plugin-local-service/commit/5251393e1065da85883d8a14eb37202147807474))
* **deps-dev:** bump @tsconfig/svelte from 4.0.1 to 5.0.2 ([576750c](https://github.com/terwer/siyuan-plugin-local-service/commit/576750c24f419f36fd6d138234b52cd2595fec0f))
* **deps-dev:** bump @types/minimist from 1.2.2 to 1.2.3 ([90bacb1](https://github.com/terwer/siyuan-plugin-local-service/commit/90bacb165259ccdf959ea0b1199ebcbc50cd4864))
* **deps-dev:** bump siyuan from 0.7.9 to 0.8.5 ([0eb4b0d](https://github.com/terwer/siyuan-plugin-local-service/commit/0eb4b0d9b561fc42d7e3bd27e7c17bb753cfff5c))
* **deps-dev:** bump stylus from 0.59.0 to 0.60.0 ([b8f0a9a](https://github.com/terwer/siyuan-plugin-local-service/commit/b8f0a9abe4c02965e112631e9a4abb0bc711fd21))
* **deps-dev:** bump svelte from 3.59.2 to 4.2.1 ([f61a71a](https://github.com/terwer/siyuan-plugin-local-service/commit/f61a71a7847e5ad465d400331542097684c5f244))
* **deps-dev:** bump vite-plugin-dts from 2.3.0 to 3.6.0 ([34d89fc](https://github.com/terwer/siyuan-plugin-local-service/commit/34d89fc8296efa4469ca6e06e2d687805c24ebd5))
* **deps-dev:** bump vite-plugin-node-polyfills from 0.8.2 to 0.15.0 ([741d0f2](https://github.com/terwer/siyuan-plugin-local-service/commit/741d0f2f3951748869d2ae8a8c537faeeb3a50c1))
* **deps-dev:** bump vite-plugin-static-copy from 0.16.0 to 0.17.0 ([12e43b6](https://github.com/terwer/siyuan-plugin-local-service/commit/12e43b6cc4a955aabecf32cdfac6f8ba5945338d))

public/app.js
module.exports = {
  dependencies: {
    core: [
      {
        name: "zhi-infra",
        version: "0.15.1",
        libpath: "libs/zhi-infra/index.cjs",
        baseType: "ThisPlugin",
        format: "cjs",
        importType: "require",
        runAs: ["Siyuan_RendererWindow"],
        initParams: ["[thisPluginBasePath]", true],
        order: 1,
      },
      {
        name: "package-downloader",
        alias: "pd",
        version: "0.0.1",
        libpath: "services/package-downloader/index.cjs",
        baseType: "ThisPlugin",
        format: "cjs",
        importType: "require",
        runAs: ["Siyuan_RendererWindow"],
        initParams: [],
        order: 2,
      },
      {
        name: "zhi-rubick-ui",
        alias: "store",
        version: "0.1.1",
        libpath: "libs/zhi-rubick-ui/index.cjs",
        baseType: "ThisPlugin",
        format: "cjs",
        importType: "require",
        runAs: ["Siyuan_RendererWindow"],
        initParams: [],
        order: 3,
      },
    ],
    server: [
      {
        name: "local-service-weekly-log",
        version: "0.0.1",
        libpath: "services/local-service-weekly-log/index.py",
        baseType: "ThisPlugin",
        format: "py",
        importType: "python",
        runAs: ["Siyuan_RendererWindow"],
        initParams: [],
        order: 101,
      },
      {
        name: "local-service-chatgpt",
        version: "0.0.1",
        libpath: "local-service-chatgpt/server.js",
        baseType: "AppService",
        format: "js",
        importType: "node",
        runAs: ["Siyuan_RendererWindow"],
        initParams: ["PORT=3000"],
        order: 102,
      },
      {
        name: "local-service-nocodb",
        version: "0.0.1",
        libpath: "local-service-nocodb/index.js",
        baseType: "AppService",
        format: "js",
        importType: "node",
        runAs: ["Siyuan_RendererWindow"],
        initParams: [
          "PORT=8000",
          "NC_DB=sqlite3:///?database[eq][siyuanDataDir]/storage/services/[thisServiceName]/noco.db",
        ],
        order: 103,
      },
    ],
    web: [],
    vendor: [],
  },
}
